src/Rules/Actions.hs

```diff
@@ -126,6 +128,18 @@ copyDirectory source target = do
     putProgressInfo $ renderAction "Copy directory" source target
     quietly $ cmd cmdEcho ["cp", "-r", source, target]
 
+-- | Copy the content of the source directory into the target directory. Only
+-- the files and directories for which the predicate returns True are copied.
+copyDirectoryContent :: (FilePath -> IO Bool) -> FilePath -> FilePath -> Action ()
+copyDirectoryContent test source target = do
+    putProgressInfo $ renderAction "Copy directory" source target
+    liftIO $ X.listFilesInside test' source >>= mapM_ cp
+  where
+    target' a = target -/- fromJust (stripPrefix source a)
+    test' a = ifM (test a) (mkdir a >> return True) (return False)
+    mkdir a = IO.createDirectoryIfMissing True $ target' a
+    cp a = whenM (test a) $ IO.copyFile a $ target' a
```

Comment:
By using system `copyFile` function you are not tracking the files being copied, that is, if they change and you rerun the `sdist` rule, Hadrian will fail to copy updated files. We either need to add `need` here, or use a tracking version of `copyFile`. However, there is an even deeper problem: we also need to depend on the result of `listFilesInside`, because some files may be added or deleted! I don't think this is currently tracked.

I think the right way to do something like this in Shake is to use Shake's `getDirectoryFiles` function. However, the problem may be that `getDirectoryFiles` does not provide a way to exclude certain files and/or directories... Maybe it should!